ComparisonWithOrleans.md

```diff
@@ -0,0 +1,203 @@
+  
+  * In Akka the central abstraction for Location Transparency is the ActorRef: it is a self-contained serializable representation of an Actor location that enables other Actors to send messages to it. ActorRefs can also be sent as part of messages to introduce Actors to one another.
+  
+##### Automatic Scale-Out
+
+  * Orleans Grains can be configured as *stateless worker* which allows multiple activations to be active at any given time. The runtime manages the scaling factor in response to the workload.
+  
+  * In Akka cluster-aware routers offer the same functionality as the stateless worker mode of Orleans with the difference that here the provided abstraction exposes available resources to the program as the cluster grows but Akka itself does not contain the logic or ability to spin up additional nodes in response to increased load. This is left to other tools that manage server fleets.
+  
+#### Actor Interfaces
+
+  * The API provided by Orleans uses code generation in order to present Actor interactions exactly like normal method invocations—with the only difference that all methods are required to return their result asynchronously. Composition of results is performed using the async–await language features of C# and friends.
+  
+  * Akka distinguishes Actor interactions from normal method invocations by requiring the use of the ActorRef’s `!` or `tell` methods in order to send messages. These method return no result value, receiving a reply from the Actor entails providing an ActorRef where the response is sent to.
+
+In this area the differences between Akka and Orleans are very pronounced and mirror the different intent stated up-front: while Orleans aims for convenience and uses syntax that is familiar to OOP practitioners, Akka makes messaging very explicit and requires the user to define message classes. There is one instance of providing a similar model as async–await, namely the interaction between a PersistentActor and its Journal using the `persist` and `persistAsync` methods, but this API leads to a stronger coupling of the two components than would otherwise be the case. For a PersistentActor this is intended since it cannot function without the Journal, but for Grain interactions users now have to choose between blocking the grain in an `await` call (in a logical sense, but still!) or permitting other invocations to be processed concurrently “during” the `await` call—which leads to much of the same race conditions and problems that shared-state concurrency has. In Akka the interaction of multiple Actors is clearly demarcated by sending messages, and if and Actor ignores all incoming requests until a certain reply has been received then that is an explicit choice that the programmer makes.
```

Comment:
If the grain code throws an error on processing application request, it is considered an error and propagated back to the client.
If the server hosting the grain dies (which I call "the service offered by the grain has failed"), the runtime automatically recovers.

I agree that there might be other types of failures or errors which you may categorize as failures and me as errors. This is a rather mute argument - the boundary is not 100% clear, right? So I don't see why this distinction is so super important. At the end we need to look at all possible/interesting errors/failures and see how we handle them. And in that regard, Orleans handles an important and a major part of failures automatically (but maybe not all possible imaginable failures).